src/main/kotlin/org/jetbrains/research/tasktracker/tracking/activity/ActivityTracker.kt

```diff
@@ -0,0 +1,127 @@
+    private val messageBusConnections: MutableList<MessageBusConnection>
+    private var trackingDisposable: Disposable? = null
+
+    init {
+        activityLogger = ActivityLogger(project)
+        messageBusConnections = mutableListOf()
+    }
+
+    // TODO: add config to select activities to track
+    fun startTracking() {
+        trackingDisposable = Disposer.newDisposable()
+        listenActions()
+        listenKeyboard()
+        listenExecution()
+    }
+
+    fun stopTracking() {
+        messageBusConnections.forEach { it.disconnect() }
+        trackingDisposable?.let { Disposer.dispose(it) }
+    }
+
+    private fun listenActions(listenShortcut: Boolean = true) {
+        val actionListener = object : AnActionListener {
+            override fun beforeActionPerformed(anAction: AnAction, event: AnActionEvent) {
+                val actionId = ActionManager.getInstance().getId(anAction) ?: return
+                if (listenShortcut && !movingActions.contains(actionId) && !typingActions.contains(actionId)) {
+                    val input = event.inputEvent
+                    if (input is KeyEvent) {
+                        activityLogger.log(Type.Shortcut, eventToShortcutInfo(input))
+                    }
+                }
+                activityLogger.log(Type.Action, actionId)
+            }
+        }
+        val connection = ApplicationManager.getApplication().messageBus.connect()
+        messageBusConnections.add(connection)
+        connection.subscribe(AnActionListener.TOPIC, actionListener)
+    }
+
+    private fun eventToShortcutInfo(event: KeyEvent): String {
+        val keys = mutableListOf<String>()
+        event.apply {
+            if (isMetaDown) {
+                keys.add("Meta")
+            }
+            if (isControlDown) {
+                keys.add("Control")
+            }
+            if (isAltDown) {
+                keys.add("Alt")
+            }
+            if (isShiftDown) {
+                keys.add("Shift")
+            }
+        }
+        keys.add(event.keyCode.toString())
+        return keys.joinToString(" ")
+    }
+
+    private fun listenExecution() {
+        val executionListener = object : ExecutionListener {
+            override fun processStarting(executorId: String, env: ExecutionEnvironment, handler: ProcessHandler) {
+                val commandLine = (handler as? BaseProcessHandler<*>)?.commandLine ?: ""
+                activityLogger.log(Type.Execution, "$executorId:'${env.runProfile}':$commandLine")
+            }
+        }
+        val connection = ApplicationManager.getApplication().messageBus.connect()
+        messageBusConnections.add(connection)
+        connection.subscribe(ExecutionManager.EXECUTION_TOPIC, executionListener)
+    }
+
+    private fun listenKeyboard() {
+        IdeEventQueue.getInstance().addPostprocessor({ awtEvent: AWTEvent ->
+            if (awtEvent is KeyEvent && awtEvent.id == KeyEvent.KEY_PRESSED) {
```

Comment:
Also, you can rewrite it a little to avoid double smart casting:
```kotlin
if (awtEvent is KeyEvent) {
    when(awtEvent.id) {
       KeyEvent.KEY_PRESSED -> ...
      KeyEvent.KEY_RELEASED -> ....
 }
}
```